### PR TITLE
Foodtype Rework (reagents now effect foodtype)

### LINKED ISSUE #43797: Any beverage in soda cans gives you mood buff.
Reproduction:
1) buy a drink from vending machine, I used star-kist as example.
2) open it and put it into the chem dispenser.
3) empty it there, mix any chemical you want inside that soda can.
4) drink the healing chemical alongside getting positive mood buff.

### PR DESCRIPTION
**Pull request by blessedmulligan**
## About The Pull Request
Foodtype (for the purposes of determining whether or not you like a certain food) now includes the reagents in that food. Drinking pure reagents will also trigger this check.
Most reagents have one or more foodtypes.
Added new foodtype: HERBAL, for use for things like coffee, tea, gin, weed, absinthe, mint, etc. Humans like it, moths dislike it (a lot of interesting plant chemicals, including caffeine and nicotine, are there as pesticides)
Containers do not have foodtypes unless the container is meant to be eaten (food are containers).
fixes #43797.

This is effectively a balance PR since it makes sweeping changes to what food and drinks are palatable. I have put exactly 0 thought or effort into what this buffs or nerfs (well, I know this nerfs moths. fuck moths).
The assignment of foodtypes is essentially arbitrary and I am more than open to suggestions.

I was a little worried this would break existing junkfood design, but I was happy to see that junkfood design and code is already shit. I'd like to cover that in a future PR.

## Why It's Good For The Game
Intuitively, the chemicals in your food should change how it tastes. This matches that intuition.
Intuitively, the type of container you drink out of should not affect how the drink tastes (see #43797). This matches that intuition.
The VEGETABLES foodtype was spread a little thin, covering everything from Kahlua to tomatoes to weed. HERBAL helps relieve some of that tension. 

## Changelog
:cl:
balance: The reagents in your food will now affect how that food tastes. This results in changes across the board.
add: Moths will find that certain herbal flavors disagree with them.
/:cl:

### CODE REVIEW

File: code/modules/reagents/reagent_containers.dm [hunk 1]
```diff
@@ -76,6 +87,34 @@
...
+/obj/item/reagent_containers/proc/checkLiked(var/fraction, mob/M)
+	if(last_check_time + 50 < world.time)
+		if(ishuman(M))
+			var/mob/living/carbon/human/H = M
+			var/total_foodtype = foodtype
+			if(reagents.reagent_list)
+				for(var/reagent in reagents.reagent_list)
+					var/datum/reagent/T = reagent
+					total_foodtype |= T.foodtype
+			if(!H.has_trait(TRAIT_AGEUSIA))
```
Comment (blessedmulligan) [inline code/modules/reagents/reagent_containers.dm]: no idea why this doesn't compile here

**Comment by actioninja** [comment]
> (well, I know this nerfs moths. fuck moths).

based and mothballed

File: code/modules/reagents/reagent_containers.dm (hunk 1 above)
Comment (actioninja) [reply code/modules/reagents/reagent_containers.dm]: has_trait was removed in favor of a macro, `HAS_TRAIT`

Comment (blessedmulligan) [reply code/modules/reagents/reagent_containers.dm]: so it has

**Comment by Shadowflame909** [comment]
Traitor chef chem kit gimmick insanely nerfed

File: code/modules/reagents/reagent_containers.dm [hunk 2]
```diff
@@ -76,6 +87,34 @@
 	. = ..()
 	SplashReagents(hit_atom, TRUE)
 
+/obj/item/reagent_containers/proc/checkLiked(var/fraction, mob/M)
```
Comment (optimumtact) [inline code/modules/reagents/reagent_containers.dm]: look at the state of this CODE LMAO

**Comment by blessedmulligan** [comment]
> 
> 
> Traitor chef chem kit gimmick insanely nerfed

My logic for adding foodtypes to poisons was more or less "if it has an obvious flavor message already, add GROSS or TOXIC or whatever, otherwise leave it NONE"

File: code/modules/reagents/reagent_containers.dm (hunk 2 above)
Comment (blessedmulligan) [reply code/modules/reagents/reagent_containers.dm]: put me out of my misery

File: code/modules/reagents/reagent_containers.dm [hunk 3]
```diff
@@ -76,6 +87,34 @@
...
+					H.adjust_disgust(11 + 15 * fraction)
+					SEND_SIGNAL(H, COMSIG_ADD_MOOD_EVENT, "gross_food", /datum/mood_event/gross_food)
+				else if(total_foodtype & H.dna.species.liked_food)
+					to_chat(H,"<span class='notice'>I love this taste!</span>")
+					H.adjust_disgust(-5 + -2.5 * fraction)
+					SEND_SIGNAL(H, COMSIG_ADD_MOOD_EVENT, "fav_food", /datum/mood_event/favorite_food)
+			else
+				if(total_foodtype & H.dna.species.toxic_food)
+					to_chat(H, "<span class='warning'>You don't feel so good...</span>")
+					H.adjust_disgust(25 + 30 * fraction)
```
Comment (MrDoomBringer) [inline code/modules/reagents/reagent_containers.dm]:
```suggestion
					H.dust()
```

**Review by MrDoomBringer — CHANGES_REQUESTED** [review]

**Comment by RaveRadbury** [comment]
Why is a soy latte herbal though

**Comment by GuyonBroadway** [comment]
So now moths hate tea eh?

British empire 2: fuck you Johnny mothboy boogaloo.

**Comment by blessedmulligan** [comment]
> 
> 
> Why is a soy latte herbal though

all coffee is labeled as herbal. it's a stretch I know

**Comment by MrPerson** [comment]
Coffee should be in its own category tbh

**Comment by Tlaltecuhtli** [comment]
how about adding COFFEE and put all the coffee subtypes there

**Comment by Jalleo** [comment]
Coffee should be herbal and energy drink.

**Comment by Mickyan** [comment]
does that mean you can use condiment packets to mask bad tastes or how does it work when you mix stuff

**Comment by Qustinnus** [comment]
nice, a refactor of my first TG pr! I always, wanted to do something like this. Maybe you can, make certain regeants drown out other tastes like condiments, which prevents dislike but not toxicity

**Comment by blessedmulligan** [comment]
> 
> 
> does that mean you can use condiment packets to mask bad tastes or how does it work when you mix stuff

bad tastes override good tastes. Condiments are pretty good now since you can add 1u of a favorite substance to any food you don't actively dislike, but I think that's better than condiments being useless.



> 
> 
> nice, a refactor of my first TG pr! I always, wanted to do something like this. Maybe you can, make certain regeants drown out other tastes like condiments, which prevents dislike but not toxicity

This is something I'd like to work on in the future, probably alongside a more intuitive junkfood system.

**Comment by blessedmulligan** [comment]
> 
> 
> how about adding COFFEE and put all the coffee subtypes there



> 
> 
> Coffee should be in its own category tbh

I'm not sure if it's best to have a foodtype for essentially one type of food (PINEAPPLE works because memes). How would you suggest species interact with COFFEE?

**Comment by Barhandar** [comment]
Enough of a taste should override other tastes as opposed to defaulting to "lol bad taste trumps everything" so that you can eat bad food if you fry and salt it enough.

**Comment by ExcessiveUseOfCobblestone** [comment]
when you fix conflicts ill take a look at it

**Pull request closed without merging**